pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/BitSetRecyclable.java

```diff
@@ -1197,11 +1199,13 @@ private BitSetRecyclable(Handle<BitSetRecyclable> recyclerHandle) {
     }
 
     public static BitSetRecyclable create() {
-        return RECYCLER.get();
+        BitSetRecyclable instance = RECYCLER.get();
+        instance.recycled.set(false);
+        return instance;
     }
 
     public void recycle() {
-        if (recyclerHandle != null) {
+        if (recyclerHandle != null && recycled.compareAndSet(false, true)) {
```

Comment:
This is still not safe. There cannot be a race condition to resolve the recycling. 

In this case it can happen: 
 1. `thread-1` calls recycle and CAS succeeds
 2. `thread-2` gets a the object from pool
 3. `thread-3` sees `recycled==false` and does the CAS again, recycle a different object that is in use. 

You cannot use Recyclable for objects with a shared ownership. Either change the logic to avoid race conditions, or use ref-counting.

Reply:
Make sense. Thanks.

Reply:
@merlimat I have addressed your comments, please take a look again, thanks.